planner/core/logical_plan_builder.go

```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
 	return joinPlan, nil
 }
 
+// rewriteFullJoin rewrites the full join to left join + anit join with null supplement.
+// select ... from t1 full join t2 on ONCOND => select ... from (select t1 left join t2 on ONCOND union all select null, null, ..., t2.* from t2 where not exists (select 1 from t1 where ONCOND))
+func (b *PlanBuilder) rewriteFullJoin(ctx context.Context, joinNode *ast.Join) (LogicalPlan, error) {
+	// build the part of left join.
+	lChild1, err := b.buildResultSetNode(ctx, joinNode.Left, false)
+	if err != nil {
+		return nil, err
+	}
+	rChild1, err := b.buildResultSetNode(ctx, joinNode.Right, false)
+	if err != nil {
+		return nil, err
+	}
+
+	handleMap1 := b.handleHelper.popMap()
+	handleMap2 := b.handleHelper.popMap()
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
+
+	leftJoinPlan := LogicalJoin{StraightJoin: joinNode.StraightJoin || b.inStraightJoin}.Init(b.ctx, b.getSelectOffset())
+	leftJoinPlan.SetChildren(lChild1, rChild1)
+	leftJoinPlan.SetSchema(expression.MergeSchema(lChild1.Schema(), rChild1.Schema()))
+	leftJoinPlan.names = make([]*types.FieldName, lChild1.Schema().Len()+rChild1.Schema().Len())
+	copy(leftJoinPlan.names, lChild1.OutputNames())
+	copy(leftJoinPlan.names[lChild1.Schema().Len():], rChild1.OutputNames())
+
+	leftJoinPlan.JoinType = LeftOuterJoin
+
+	// Merge sub-plan's fullSchema into this join plan.
+	// Please read the comment of LogicalJoin.fullSchema for the details.
+	var (
+		lFullSchema, rFullSchema *expression.Schema
+		lFullNames, rFullNames   types.NameSlice
+	)
+	if left, ok := lChild1.(*LogicalJoin); ok && left.fullSchema != nil {
+		lFullSchema = left.fullSchema
+		lFullNames = left.fullNames
+	} else {
+		lFullSchema = lChild1.Schema()
+		lFullNames = lChild1.OutputNames()
+	}
+	if right, ok := rChild1.(*LogicalJoin); ok && right.fullSchema != nil {
+		rFullSchema = right.fullSchema
+		rFullNames = right.fullNames
+	} else {
+		rFullSchema = rChild1.Schema()
+		rFullNames = rChild1.OutputNames()
+	}
+	if joinNode.Tp == ast.RightJoin {
+		// Make sure lFullSchema means outer full schema and rFullSchema means inner full schema.
+		lFullSchema, rFullSchema = rFullSchema, lFullSchema
+		lFullNames, rFullNames = rFullNames, lFullNames
+	}
+	leftJoinPlan.fullSchema = expression.MergeSchema(lFullSchema, rFullSchema)
+
+	// Merge sub-plan's fullNames into this join plan, similar to the fullSchema logic above.
+	leftJoinPlan.fullNames = make([]*types.FieldName, 0, len(lFullNames)+len(rFullNames))
+	for _, lName := range lFullNames {
+		name := *lName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+	for _, rName := range rFullNames {
+		name := *rName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+
+	// Clear NotNull flag for the inner side schema if it's an outer join.
+	if joinNode.Tp == ast.LeftJoin || joinNode.Tp == ast.RightJoin {
+		resetNotNullFlag(leftJoinPlan.fullSchema, 0, leftJoinPlan.fullSchema.Len())
+	}
```

Comment:
The `joinNode.Tp` must be `FullJoin`, but we need to treat the logical join as a `LeftJoin` here.